src/DeviceInput/deviceInputSystem.ts

```diff
@@ -0,0 +1,225 @@
+            if (pointerY) {
+                pointerY[1] = evt.clientY;
+            }
+        });
+
+        this._elementToAttachTo?.addEventListener("pointerdown", (evt) => {
+            if (!this._pointerActive) {
+                this._pointerActive = true;
+                this.registerDevice(DeviceInputSystem.POINTER_DEVICE, this._maxPointerButtons);
+                this._onDeviceConnected(DeviceInputSystem.POINTER_DEVICE);
+            }
+
+            let pointerButton = this._inputs.get(DeviceInputSystem.POINTER_DEVICE);
+            if (pointerButton) {
+                pointerButton[evt.button] = 1;
+            }
+        });
+
+        this._elementToAttachTo?.addEventListener("pointerup", (evt) => {
+            let pointerButton = this._inputs.get(DeviceInputSystem.POINTER_DEVICE);
+            if (pointerButton) {
+                pointerButton[evt.button] = 0;
+            }
+        });
+    }
+
+    /**
+     * Handle all actions that come from gamepad interaction
+     */
+    private handleGamepadActions() {
+        window.addEventListener("gamepadconnected", (evt: any) => {
+            var deviceName = `${evt.gamepad.id}-${evt.gamepad.index}`;
+            this.registerDevice(deviceName, evt.gamepad.buttons.length + evt.gamepad.axes.length);
+            this._onDeviceConnected(deviceName);
+        });
+
+        window.addEventListener("gamepaddisconnected", (evt: any) => {
+            var deviceName = `${evt.gamepad.id}-${evt.gamepad.index}`;
+            this.deregisterDevice(deviceName);
+            this._onDeviceDisconnected(deviceName);
+        });
+    }
+
+    /**
+     * Update all non-event based devices with each frame
+     */
+    private updateDevices() {
+        // Gamepads
+        var gamepads = this.getGamePads();
+
+        for (var j = 0; j < gamepads.length; j++) {
+            let gp = gamepads[j];
+
+            if (gp) {
+                for (var i = 0; i < gp.buttons.length; i++) {
+                    let button = this._inputs.get(`${gp.id}-${gp.index}`);
+                    if (button) {
+                        button[i] = gp.buttons[i].value;
+                    }
+                }
+
+                for (var i = 0; i < gp.axes.length; i++) {
+                    let axis = this._inputs.get(`${gp.id}-${gp.index}`);
+                    if (axis) {
+                        axis[i + gp.buttons.length] = gp.axes[i].valueOf();
+                    }
+                }
+            }
+        }
+
+        Engine.QueueNewFrame(() => { this.updateDevices(); });
+    }
+
+    /**
+     * getGamePads: returns all gamepads
+     * @returns array with active gamepads
+     */
+    private getGamePads(): (Gamepad | null)[] {
+        return navigator.getGamepads();
+    }
+
+    // Static
+    /** POINTER_DEVICE */
+    public static readonly POINTER_DEVICE: string = "Pointer";
+    /** KEYBOARD_DEVICE */
+    public static readonly KEYBOARD_DEVICE: string = "Keyboard";

```

Comment:
I recommend putting things that are part of the contract of this class together at the top.

Reply:
I put this at the bottom to match the layout of other BJS files.  Personally, I prefer it at the top so if it's all the same, I'll move this back to the top.

Reply:
Agree with @bghgary, it's weird seeing these at the bottom.

Reply:
The majority of the code puts these statics at the top from what I can see.